code/datums/proximity/advanced_proximity_monitor.dm

```diff
@@ -0,0 +1,167 @@
+	var/list/turf/edge_turfs = list()
+
+/datum/proximity_monitor/advanced/Destroy()
+	cleanup_field()
+	return ..()
+
+/datum/proximity_monitor/advanced/proc/cleanup_field()
+	for(var/turf/turf as anything in edge_turfs)
+		cleanup_edge_turf(turf)
+	edge_turfs = list()
+	for(var/turf/turf as anything in field_turfs)
+		cleanup_field_turf(turf)
+	field_turfs = list()
+
+//Call every time the field moves (done automatically if you use update_center) or a setup specification is changed.
+/datum/proximity_monitor/advanced/proc/recalculate_field(full_recalc = FALSE)
+	var/list/new_turfs = update_new_turfs()
+
+	var/list/old_field_turfs = field_turfs
+	var/list/old_edge_turfs = edge_turfs
+	field_turfs = new_turfs[FIELD_TURFS_KEY]
+	edge_turfs = new_turfs[EDGE_TURFS_KEY]
+	if(full_recalc)
+		field_turfs = list()
+		edge_turfs = list()
+
+	for(var/turf/old_turf as anything in old_field_turfs - field_turfs)
+		if(QDELETED(src))
+			return
+		cleanup_field_turf(old_turf)
+	for(var/turf/old_turf as anything in old_edge_turfs - edge_turfs)
+		if(QDELETED(src))
+			return
+		cleanup_edge_turf(old_turf)
+
+	if(full_recalc)
+		old_field_turfs = list()
+		old_edge_turfs = list()
+		field_turfs = new_turfs[FIELD_TURFS_KEY]
+		edge_turfs = new_turfs[EDGE_TURFS_KEY]
+
+	for(var/turf/new_turf as anything in field_turfs - old_field_turfs)
+		if(QDELETED(src))
+			return
+		setup_field_turf(new_turf)
+
+	for(var/turf/new_turf as anything in edge_turfs - old_edge_turfs)
+		if(QDELETED(src))
+			return
+		setup_edge_turf(new_turf)
+
+/datum/proximity_monitor/advanced/on_initialized(turf/location, atom/created, init_flags)
+	. = ..()
+	on_entered(location, created, null)
+
+/datum/proximity_monitor/advanced/on_entered(turf/source, atom/movable/entered, turf/old_loc)
+	. = ..()
+	if(get_dist(source, host) == current_range)
+		field_edge_crossed(entered, old_loc, source)
+	else
+		field_turf_crossed(entered, old_loc, source)
+
+/datum/proximity_monitor/advanced/on_moved(atom/movable/movable, atom/old_loc)
+	. = ..()
+	if(ignore_if_not_on_turf)
+		//Early return if it's not the host that has moved.
+		if(movable != host)
+			return
+		//Cleanup the field if the host was on a turf but isn't anymore.
+		if(!isturf(host.loc))
+			if(isturf(old_loc))
+				cleanup_field()
+			return
+	recalculate_field(full_recalc = FALSE)
+
+/datum/proximity_monitor/advanced/on_uncrossed(turf/source, atom/movable/gone, direction)
+	if(get_dist(source, host) == current_range)
+		field_edge_uncrossed(gone, source, get_turf(gone))
+	else
+		field_turf_uncrossed(gone, source, get_turf(gone))
+
+/// Called when a turf in the field of the monitor is linked
+/datum/proximity_monitor/advanced/proc/setup_field_turf(turf/target)
+	return
+
+/// Called when a turf in the field of the monitor is unlinked
+/// Do NOT call this manually, requires management of the field_turfs list
+/datum/proximity_monitor/advanced/proc/cleanup_field_turf(turf/target)
```

Comment:
You could probably also use some of the encapsulation helpers for this to mark it private